attn_gradcheck.py

```diff
@@ -0,0 +1,45 @@
+import torch
+
+class Attn(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, q, k, v):
+        attn = torch.matmul(q, k.swapaxes(0, 1))
+        attn = torch.tanh(attn)
+        output = torch.matmul(attn, v)
+
+        ctx.save_for_backward(q, k, v, attn)
+        return output, attn
+
+    @staticmethod
+    def backward(ctx, grad_output, grad_attn):
+        (q, k, v, attn) = ctx.saved_tensors
+
+        # computes and stores sech^2(qk')
+        partial = 1 / torch.cosh(q.matmul(k.swapaxes(0, 1))).pow(2)
+
+        # computes grad_output terms
+        grad_q = grad_output.matmul(v.swapaxes(0, 1)).mul(partial).matmul(k)
```

Comment:
nit: you can actually pre-compute `grad_output.matmul(v.swapaxes(0, 1)).mul(partial)` and re-use it here.